docs/arch/003-testing-flows.md

```diff
@@ -0,0 +1,174 @@
+
+### [Flow unit tests](https://github.com/alphagov/smart-answers/tree/93f34cae3eb746b5b08d78b6ea6b7d5c1dfbbc18/test/unit/smart_answer_flows)
+
+These tests:
+
+- are intended as a complement to Flow integration tests with individual tests for [Flow](https://github.com/alphagov/smart-answers/blob/93f34cae3eb746b5b08d78b6ea6b7d5c1dfbbc18/test/unit/smart_answer_flows/part_year_profit_tax_credits_flow_test.rb) and [ERB files](https://github.com/alphagov/smart-answers/blob/93f34cae3eb746b5b08d78b6ea6b7d5c1dfbbc18/test/unit/smart_answer_flows/part_year_profit_tax_credits_view_test.rb)
+- have moderate adoption, 10 / 36 Flows have flow tests, 3 / 36 Flows have view tests
+- duplicate the assertions between Flow unit tests and Flow integration tests
+
+### [RSpec feature tests](https://github.com/alphagov/smart-answers/tree/93f34cae3eb746b5b08d78b6ea6b7d5c1dfbbc18/spec/features/flows)
+
+These tests:
+
+- have low adoption, 5 / 36 Flows have tests
+- are built upon RSpec test framework which is [being removed from Smart Answers](https://github.com/alphagov/smart-answers/issues/5350)
+- exercise a full integration (including ERB files) by exercising and asserting via HTML output
+- are significantly slower than other test approaches - testing through Rails requests is [approximately 50x slower](https://github.com/alphagov/smart-answers/pull/5394)
+
+## Decision
+
+We have decided to adopt a new testing approach that will supersede the aforementioned approaches to testing Smart Answer Flows. The goals of this new approach are to provide:
+
+- tests that are fast to run
+- assertions made at an ERB level (to prove their execution)
+- a method to achieve full (or at least very high) coverage of Flow files
+- easy to understand tests (relatively, given how hard Smart Answers can be to understand)
+- relatively low levels of tedium/repetition in reading/writing tests (ideally length of test flow file is within 2x the flow file length)
+- testing approaches that are consistent across different flow approaches (URI path, query string and session)
+
+Following [two](https://github.com/alphagov/smart-answers/pull/5394) [rounds](https://github.com/alphagov/smart-answers/pull/5440) of feedback we've established that the new approach will:
+
+- consolidate integration and unit testing to a single approach, similar to Rails' controller testing (with a similar emphasis on integration testing), since there is high possibility of tedious levels of duplication between multiple files
+- have a boundary of testing the Flow and it's contents, explicitly not testing earlier considerations such as routing or controller - this is due to them being tested separately and the performance advantages gained from not including them
```

Comment:
You might have meant: `have a boundary of testing the Flow and its contents` as it is possessive.